app/controllers/products_controller.rb

```diff
@@ -0,0 +1,49 @@
+class ProductsController < ApplicationController
+  before_action :find_product, only: [ :show, :update, :destroy ]
+
+  def index
+    @products = Product.all
+    render json: @products
+  end
+
+  def show
+    render json: @product
+  end
+
+  def create
+    @product = Product.new(product_params)
+    if @product.save
+      render json: @product, status: :created, location: @product
+    else
+      render json: @product.errors, status: :unprocessable_entity
+    end
+  end
+
+  def update
+    if @product
+      @product.update(product_params)
+      render json: @product
+    else
+      render json: @product.errors, status: :unprocessable_entity
+    end
+  end
+
+  def destroy
+    if @product
+      @product.destroy
+      render json: { message: "Product successfully deleted." }, status: 200
+    else
+      render json: { error: "Product not found." }, status: 404
+    end
+  end
+
+  private
+
+  def product_params
+    params.require(:product).permit(:name, :price, :user_id)
+  end
+
+  def find_product
+    @product = Product.find(params[:id])
```

Comment:
Hàm find sẽ raise exception nếu ko tìm thấy.
Cần rescue để return 404 luôn, thì trong action khỏi check if/else.

Tham khảo thêm về pattern này để handle lỗi nhàn hơn.
https://viblo.asia/p/error-handling-in-rails-api-the-rails-way-RQqKLvRrl7z